CTOR-534 Refresh cloud-kubernetes-api doc (#3437)
centreon/centreon-documentation: pp/integrations/plugin-packs/procedures/cloud-kubernetes-api.md

id: cloud-kubernetes-api
title: Kubernetes API
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

## Pack assets

### Preamble

> This Pack aims to monitor both infrastructure layer (nodes) and cluster

Reviewer: ```suggestion
This Pack aims to monitor both the infrastructure layer (nodes) and the cluster
```

> services (deployments, daemonsets, etc). The Kubernetes API Pack gives 

Reviewer: ```suggestion
services (deployments, daemonsets, etc) of a Kubernetes cluster. The Kubernetes API pack gives 
```

> multiple choices regarding the way you can arrange a cluster monitoring. 

Reviewer: ```suggestion
multiple choices regarding the way you can arrange cluster monitoring. 
```

> There is mainly three ways:

Reviewer: ```suggestion
There are mainly three ways:
```

- Gather all metrics on only one Centreon host with a service per Kubernetes
  unit (i.e. deployments, daemonsets, etc) - apply
  [manual creation](#using-a-host-template-provided-by-the-connector) procedure,
- Gather all metrics on only one Centreon host with a service for each

>   instances of each Kubernetes units - apply [manual creation](#using-a-host-template-provided-by-the-connector)

Reviewer: ```suggestion
  instance of each Kubernetes unit - apply [manual creation](#using-a-host-template-provided-by-the-connector)
```

  and [service discovery](#using-a-service-template-provided-by-the-connector) procedures,
- Collect infrastructural metrics (master and worker nodes) with a Centreon
  host per Kubernetes node, and keep orchestration/application metrics

>   on a unique host (using one of the 2 previous scenarii) - apply

Reviewer: ```suggestion
  on a unique host (using one of the 2 previous scenarios) - apply the
```

  [host discovery](/docs/monitoring/discovery/hosts-discovery) procedure.

### Templates

The Monitoring Connector **Kubernetes API** brings 2 host templates:

* **Cloud-Kubernetes-Api-custom**
* **Cloud-Kubernetes-Kubectl-custom**
* **Cloud-Kubernetes-Node-Api-custom**
* **Cloud-Kubernetes-Node-Kubectl-custom**

The connector brings the following service templates (sorted by the host template they are attached to):

<Tabs groupId="sync">
<TabItem value="Cloud-Kubernetes-Api-custom" label="Cloud-Kubernetes-Api-custom">

| Service Alias                | Service Template                                         | Service Description                                                                                        | Discovery  |
|:-----------------------------|:---------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------|:----------:|

> | Cluster-Events               | Cloud-Kubernetes-Cluster-Events-Api-custom               | Check the number of events occurring on the cluster                                                        |            |

Reviewer: ```suggestion
| Cluster-Events               | Cloud-Kubernetes-Cluster-Events-Api-custom               | Check the number of events occurring in the cluster                                                        |            |
```

| CronJob-Status               | Cloud-Kubernetes-CronJob-Status-Api-custom               | Check CronJobs status                                                                                      | X          |
| Daemonset-Status             | Cloud-Kubernetes-Daemonset-Status-Api-custom             | Check DaemonSets status                                                                                    | X          |
| Deployment-Status            | Cloud-Kubernetes-Deployment-Status-Api-custom            | Check Deployments status                                                                                   | X          |
| Node-Status                  | Cloud-Kubernetes-Node-Status-Api-custom                  | Check Nodes status                                                                                         |            |

> | Node-Status                  | Cloud-Kubernetes-Node-Status-Name-Api-custom             | Checks status of a node identified by its name (for example at the end of the associated discovery rule)   | X          |

Reviewer: ```suggestion
| Node-Status                  | Cloud-Kubernetes-Node-Status-Name-Api-custom             | Check status of a node identified by its name (for example at the end of the associated discovery rule)   | X          |
```

| ReplicationController-Status | Cloud-Kubernetes-ReplicationController-Status-Api-custom | Check ReplicationControllers status                                                                        | X          |
| StatefulSet-Status           | Cloud-Kubernetes-StatefulSet-Status-Api-custom           | Check StatefulSets status                                                                                  | X          |

> The services listed above are created automatically when the **Cloud-Kubernetes-Api-custom** host template is used.

> If **Discovery** is checked, it means a service discovery rule exists for this service template.

</TabItem>
<TabItem value="Cloud-Kubernetes-Kubectl-custom" label="Cloud-Kubernetes-Kubectl-custom">

| Service Alias                | Service Template                                         | Service Description                                                                                        | Discovery  |
|:-----------------------------|:---------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------|:----------:|

> | Cluster-Events               | Cloud-Kubernetes-Cluster-Events-Api-custom               | Check the number of events occurring on the cluster                                                        |            |

Reviewer: ```suggestion
| Cluster-Events               | Cloud-Kubernetes-Cluster-Events-Api-custom               | Check the number of events occurring in the cluster                                                        |            |
```

| CronJob-Status               | Cloud-Kubernetes-CronJob-Status-Api-custom               | Check CronJobs status                                                                                      | X          |
| Daemonset-Status             | Cloud-Kubernetes-Daemonset-Status-Api-custom             | Check DaemonSets status                                                                                    | X          |
| Deployment-Status            | Cloud-Kubernetes-Deployment-Status-Api-custom            | Check Deployments status                                                                                   | X          |
| Node-Status                  | Cloud-Kubernetes-Node-Status-Api-custom                  | Check Nodes status                                                                                         |            |

> | Node-Status                  | Cloud-Kubernetes-Node-Status-Name-Api-custom             | Checks status of a node identified by its name (for example at the end of the associated discovery rule)   | X          |

Reviewer: ```suggestion
| Node-Status                  | Cloud-Kubernetes-Node-Status-Name-Api-custom             | Check the status of a node identified by its name (for example at the end of the associated discovery rule)   | X          |
```

| StatefulSet-Status           | Cloud-Kubernetes-StatefulSet-Status-Api-custom           | Check StatefulSets status                                                                                  | X          |

> The services listed above are created automatically when the **Cloud-Kubernetes-Kubectl-custom** host template is used.

> If **Discovery** is checked, it means a service discovery rule exists for this service template.

</TabItem>
<TabItem value="Cloud-Kubernetes-Node-Api-custom" label="Cloud-Kubernetes-Node-Api-custom">

| Service Alias | Service Template                             | Service Description                                                                                        | Discovery  |
|:--------------|:---------------------------------------------|:-----------------------------------------------------------------------------------------------------------|:----------:|
| Node-Status   | Cloud-Kubernetes-Node-Status-Api-custom      | Check Nodes status                                                                                         |            |

> | Node-Status   | Cloud-Kubernetes-Node-Status-Name-Api-custom | Checks status of a node identified by its name (for example at the end of the associated discovery rule)   | X          |

Reviewer: ```suggestion
| Node-Status   | Cloud-Kubernetes-Node-Status-Name-Api-custom | Check the status of a node identified by its name (for example at the end of the associated discovery rule)   | X          |
```

| Node-Usage    | Cloud-Kubernetes-Node-Usage-Name-Api-custom  | Check the usage of a node identified by its name (for example at the end of the associated discovery rule) | X          |

> The services listed above are created automatically when the **Cloud-Kubernetes-Node-Api-custom** host template is used.

> If **Discovery** is checked, it means a service discovery rule exists for this service template.

</TabItem>
<TabItem value="Cloud-Kubernetes-Node-Kubectl-custom" label="Cloud-Kubernetes-Node-Kubectl-custom">

| Service Alias | Service Template                             | Service Description                                                                                        | Discovery  |
|:--------------|:---------------------------------------------|:-----------------------------------------------------------------------------------------------------------|:----------:|
| Node-Status   | Cloud-Kubernetes-Node-Status-Api-custom      | Check Nodes status                                                                                         |            |

> | Node-Status   | Cloud-Kubernetes-Node-Status-Name-Api-custom | Checks status of a node identified by its name (for example at the end of the associated discovery rule)   | X          |

Reviewer: ```suggestion
| Node-Status   | Cloud-Kubernetes-Node-Status-Name-Api-custom | Check the status of a node identified by its name (for example at the end of the associated discovery rule)   | X          |
```

| Node-Usage    | Cloud-Kubernetes-Node-Usage-Api-custom       | Check nodes usage                                                                                          |            |
| Node-Usage    | Cloud-Kubernetes-Node-Usage-Name-Api-custom  | Check the usage of a node identified by its name (for example at the end of the associated discovery rule) | X          |

> The services listed above are created automatically when the **Cloud-Kubernetes-Node-Kubectl-custom** host template is used.

> If **Discovery** is checked, it means a service discovery rule exists for this service template.

</TabItem>
</Tabs>

### Discovery rules

#### Host discovery

| Rule name                  | Description                                                              |
|----------------------------|--------------------------------------------------------------------------|

> | Kubernetes Nodes (RestAPI) | Discover Kubernetes nodes by requesting Kubernetes RestAPI               |

Reviewer: ```suggestion
| Kubernetes Nodes (RestAPI) | Discover Kubernetes nodes by requesting the Kubernetes RestAPI               |
```

> | Kubernetes Nodes (Kubectl) | Discover Kubernetes nodes by requesting Kubernetes cluster using kubectl |

Reviewer: ```suggestion
| Kubernetes Nodes (Kubectl) | Discover Kubernetes nodes by requesting the Kubernetes cluster using kubectl |
```

| *pods*~pod-status                             | N/A   |
| *pods*~restarts.total.count                   | count |
| *pods*~*containers*#container-status          | N/A   |
| *pods*~*containers*#containers.restarts.count | count |

</TabItem>
<TabItem value="ReplicaSet-Status" label="ReplicaSet-Status">

| Metric name                                   | Unit  |
|:----------------------------------------------|:------|
| *replicasets*#replicaset.replicas.ready.count | count |

</TabItem>
<TabItem value="ReplicationController-Status" label="ReplicationController-Status">

| Metric name                                      | Unit  |
|:-------------------------------------------------|:------|
| *rcs*#replicationcontroller.replicas.ready.count | count |

</TabItem>
<TabItem value="StatefulSet-Status" label="StatefulSet-Status">

| Metric name                                     | Unit  |
|:------------------------------------------------|:------|
| *statefulsets*#statefulset.replicas.ready.count | count |

</TabItem>
</Tabs>

### Additional information on metrics and services

<Tabs groupId="sync">
<TabItem value="Cluster-Events" label="Cluster-Events">

> This indicator allows to watch the number of events occurring on the cluster,

Reviewer: ```suggestion
This indicator allows you to watch the number of events occurring on the cluster,
```

Event 'Warning' for object 'Pod/graphite-0' with message 'Liveness probe failed: Get "http://10.244.2.10:8080/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)', Count: 1, First seen: 26m 21s ago (2021-03-11T12:26:23Z), Last seen: 26m 21s ago (2021-03-11T12:26:23Z)
```

</TabItem>
<TabItem value="CronJob-Status" label="CronJob-Status">

This indicator allows to check that CronJobs are executed as they should,

> like the `kubectl get cronjobs` can provide:

Reviewer: like ... can provide

> For the host discovery: set the token [retrieved ealier](#retrieve-token-from-service-account) from the service account,

</TabItem>
<TabItem value="Cloud-Kubernetes-Kubectl-custom" label="Cloud-Kubernetes-Kubectl-custom">

1. Log into Centreon and add a new host through **Configuration > Hosts**.
2. Fill in the **Name**, **Alias** & **IP Address/DNS** fields according to your resource's settings.
3. Apply the **Cloud-Kubernetes-Kubectl-custom** template to the host. A list of macros appears. Macros allow you to define how the connector will connect to the resource, and to customize the connector's behavior.
4. Fill in the macros you want. Some macros are mandatory, in particular the macro for defining the [custom mode](#available-custom-modes), i.e. the connection method to the resource.

| Macro                   | Description                                                                                                                | Default value     | Mandatory   |
|:------------------------|:---------------------------------------------------------------------------------------------------------------------------|:------------------|:-----------:|

> | KUBECTLCONFIGFILE       | Kubernetes configuration file path (default: '~/.kube/config'). (example: --config-file='/root/.kube/config').             |                   | X           |

Reviewer 2: ```suggestion
| KUBECTLCONFIGFILE       | Kubernetes configuration file path.             | ~/.kube/config | X           |
```

| EXTRAOPTIONS            | Any extra option you may want to add to every command (a --verbose flag for example). All options are listed [here](#available-options).                       |                   |             |

5. [Deploy the configuration](/docs/monitoring/monitoring-servers/deploying-a-configuration). The host appears in the list of hosts, and on the **Resources Status** page. The command that is sent by the connector is displayed in the details panel of the host: it shows the values of the macros.
> For the host discovery: set the token [retrieved ealier](#retrieve-token-from-service-account) from the service account,
> - If using kubectl: set the path to the [created configuration file](#create-a-kubectl-configuration) (prefer using relative path to make it work for both discovery and monitoring, i.e. `~/.kube/config`).

</TabItem>
<TabItem value="Cloud-Kubernetes-Node-Kubectl-custom" label="Cloud-Kubernetes-Node-Kubectl-custom">

1. Log into Centreon and add a new host through **Configuration > Hosts**.
2. Fill in the **Name**, **Alias** & **IP Address/DNS** fields according to your resource's settings.

> 3. Apply the **Cloud-Kubernetes-Node-Api-custom** template to the host. A list of macros appears. Macros allow you to define how the connector will connect to the resource, and to customize the connector's behavior.

Author: ```suggestion
3. Apply the **Cloud-Kubernetes-Node-Kubectl-custom** template to the host. A list of macros appears. Macros allow you to define how the connector will connect to the resource, and to customize the connector's behavior.
```

4. Fill in the macros you want. Some macros are mandatory, in particular the macro for defining the [custom mode](#available-custom-modes), i.e. the connection method to the resource.

| Macro                   | Description                                                                                                                | Default value     | Mandatory   |
|:------------------------|:---------------------------------------------------------------------------------------------------------------------------|:------------------|:-----------:|

> | KUBECTLCONFIGFILE       | Kubernetes configuration file path (default: '~/.kube/config'). (example: --config-file='/root/.kube/config').             |                   | X           |

Reviewer 2: ```suggestion
| KUBECTLCONFIGFILE       | Kubernetes configuration file path.             | ~/.kube/config | X           |
```
